### PR TITLE
Fixing wrongful release of arrow arrays

### PR DESCRIPTION
**Pull request by pdet**
Since we reference pointers from arrow arrays, we must hold them until the full query execution cycle.

I've added shared_ptrs since multiple data chunks can reference the same arrow array.

@jonkeane can you check if this also solves the issue from your side?

### CODE REVIEW

**Comment by jonkeane** [comment]
I've installed this and it does, indeed seem to fix _most_ of the issues, we're getting very close!

I am noticing, though that if there is a `mutate()` after  `to_duckdb()` we're still seeing some corruption.

One example that you could add to your test file is:

```
ds <- open_dataset(rep("data/userdata1.parquet", 4))

 to_to <- ds %>%
   select(-registration_dttm) %>% # timestamp[ns] has unrelated error
   to_duckdb() %>%
   mutate(new_id = id + 1) %>%
   to_arrow() %>%
   collect() %>%
   arrange(id)

 all_arrow <- ds %>%
   select(-registration_dttm) %>% # timestamp[ns] has unrelated error
   mutate(new_id = id + 1) %>%
   collect() %>%
   arrange(id)


 # The full comparison
 expect_true(all.equal(to_to,all_arrow))
```

When I run that, I'm getting:
```
>  # The full comparison
>  expect_true(all.equal(to_to,all_arrow))
Error: all.equal(to_to, all_arrow) is not TRUE

`actual` is a character vector ('Component “id”: Mean relative difference: 0.9979156', 'Component “first_name”: 1708 string mismatches', 'Component “last_name”: 1708 string mismatches', 'Component “email”: 1723 string mismatches', 'Component “gender”: 864 string mismatches', ...)
`expected` is a logical vector (TRUE)
> to_to
# A tibble: 4,000 × 13
         id first_name last_name email gender ip_address cc    country birthdate
      <int> <chr>      <chr>     <chr> <chr>  <chr>      <chr> <chr>   <chr>    
 1  -9.64e7 Amanda     Jordan    ajor… Female 1.197.201… "675… Indone… 3/8/1971 
 2  -7.46e7 Amanda     Jordan    ajor… Female 1.197.201… "675… Indone… 3/8/1971 
 3  -6.68e7 Amanda     Jordan    ajor… Female 1.197.201… "675… Indone… 3/8/1971 
 4   0      Amanda     Jordan    ajor… Female 1.197.201… "675… Indone… 3/8/1971 
 5   0      Albert     Freeman   afre… Male   218.111.1… ""    Canada  1/16/1968
 6   3   e0 Evelyn     Morgan    emor… Female 7.161.136… "676… Russia  2/1/1960 
 7   3   e0 Evelyn     Morgan    emor… Female 7.161.136… "676… Russia  2/1/1960 
 8   3   e0 Evelyn     Morgan    emor… Female 7.161.136… "676… Russia  2/1/1960 
 9   3   e0 Evelyn     Morgan    emor… Female 7.161.136… "676… Russia  2/1/1960 
10   4   e0 Denise     Riley     dril… Female 140.35.10… "357… China   4/8/1997 
# … with 3,990 more rows, and 4 more variables: salary <dbl>, title <chr>,
#   comments <chr>, new_id <dbl>
```

In case it's helpful, here's what the SQL is that dplyr is generating:
```
> ds %>%
+    select(-registration_dttm) %>% # timestamp[ns] has unrelated error
+    to_duckdb() %>%
+    mutate(new_id = id + 1) %>%
+    show_query()
<SQL>
SELECT "id", "first_name", "last_name", "email", "gender", "ip_address", "cc", "country", "birthdate", "salary", "title", "comments", "id" + 1.0 AS "new_id"
FROM "arrow_067"
```

Might there be one more place where we need to use a shared pointer?

**Comment by pdet** [comment]
@jonkeane can you give it another try? I think this should be fixed now, just had to figure out a better place to place the arrow arrays :-)

Thanks again for checking this!

**Comment by jonkeane** [comment]
That worked for me, our tests are passing now locally!

File: src/include/duckdb/common/types/vector_buffer.hpp [hunk 1]
```diff
@@ -48,6 +50,7 @@ class VectorBuffer {
 	}
 	VectorBuffer() {
 	}
+	shared_ptr<ArrowArrayWrapper> arrow_array;
```
Comment (Mytherin) [inline src/include/duckdb/common/types/vector_buffer.hpp:53]: Could we make this more generic? I'm thinking about having a generic struct, e.g.:


```cpp
struct VectorAuxiliaryData {
   virtual ~ VectorAuxiliaryData();
};
```

The VectorBuffer can then have a `unique_ptr< VectorAuxiliaryData> extra_data;` as a member. Inside the Arrow integration, you can then inherit from the `VectorAuxiliaryData`, and create a `ArrowVectorAuxiliaryData` object.

This way we (1) don't need to include the arrow wrapper stuff everywhere the vector_buffer is included, and (2) can leave this as a more generic solution that others can also use.

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! One comment:

File: src/include/duckdb/common/types/vector_buffer.hpp (hunk 1 above)
Comment (hannesmuehleisen) [reply src/include/duckdb/common/types/vector_buffer.hpp:53]: Yes that was also my concern here

**Pull request merged**